### ISSUE TITLE
Syntax too permissive for list ?

### ISSUE THREAD

**Issue by tpoindessous** [hashicorp/terraform#15529]
Hi

I got an error when using list from datasource instead of list from variable or plaintext.

## Terraform version
Terraform v0.9.11

## Code 
```
.
├── module
│   └── main.tf
└── topology.tf
```

`cat topology.tf`
```
# example 1: working
module "security_group_admin" {
  source                             = "./module/"
  security_group_ingress_cidr_blocks = ["11.0.0.0/25", "10.0.1.0/24"]
}

/*
# example 2: not working with line module/main.tf:12 but works with line module/main.tf:15
module "security_group_admin" {
  source                             = "./module/"
  security_group_ingress_cidr_blocks = ["${data.aws_vpc.test.cidr_block}", "10.0.1.0/24"]
}
data "aws_vpc" "test" {
  id = "123456"
}
*/
```

`cat module/main.tf`
```
resource "aws_security_group" "admin" {
  name   = "admin"
  vpc_id = "123456"
}

resource "aws_security_group_rule" "admin_ingress_ssh" {
  type              = "ingress"
  from_port         = 22
  to_port           = 22
  protocol          = "tcp"
  # Works only for example1
  cidr_blocks       = "${var.security_group_ingress_cidr_blocks}"

  # Works for example1 and example2
  #cidr_blocks       = ["${var.security_group_ingress_cidr_blocks}"]
  security_group_id = "${aws_security_group.admin.id}"
}

variable "security_group_ingress_cidr_blocks" {
  type = "list"
}
```

So when I'm using list from variables (or plaintext in my example), I can use this syntax :
`cidr_blocks       = "${var.security_group_ingress_cidr_blocks}"` 
but if I want to use list from datasource (like cidr_block from aws_vpc), I must use this syntax :
`cidr_blocks       = ["${var.security_group_ingress_cidr_blocks}"]`

I'm fine with that, but I would expect at least a little warning when using the first syntax :)

Thanks !

**Comment by bcokert** [comment]
Why does it do this in the first place? Shouldn't the result be a list of a list? I have the same thing were I pass a list into a module, who's variable is a list. But to pass that list to a resource I have to wrap it in another list:

```hcl
//main.tf
module "controller_b" {
  source = "./modules/controller"
  security_group_ids = ["stringFromSomewhere"]
}

//modules/controller/main.tf
variable "security_group_ids" { type = "list" }
resource "aws_instance" "controller" {
  vpc_security_group_ids = "${var.security_group_ids}" # fails, even though it's already a list
  vpc_security_group_ids = ["${var.security_group_ids}"] #works
}
```

The error is:
```console
module.controller_b.aws_instance.controller: vpc_security_group_ids: should be a list
```

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

This issue is related to #12263, which I've closed today since it appears to be fixed in v0.12.0-alpha1.

Very early versions of Terraform did not have support for true lists at all, and so the confusing behavior you've seen here is a result of a backward-compatibility shim to allow an old syntax where list brackets were used as a hint to interpret a value as a "pseudo-list". This legacy compatibility syntax later found a new purpose as a workaround for bug #12263, which is caused by Terraform losing track of the type of an "unknown" list value during expression evaluation.

The forthcoming v0.12.0 final release both addresses #12263 and removes this long-standing compatibility behavior with extra list brackets. As a result, the redundant set of list brackets is no longer necessary and will indeed produce a list of lists as @bcokert expected in the first place.

In v0.12.0, the correct form for the example in the original comment on this issue would be:

```hcl
  cidr_blocks = var.security_group_ingress_cidr_blocks
```

Notice that the interpolation syntax is no longer required either. Retaining it won't do any harm (Terraform will automatically disregard it just like it did before), but we'll recommend using this first-class expression form in future since it is easier to read and understand.

Since it seems like the concern raised in this issue has been addressed in `master`,  I'm going to close this out. Thanks for raising it, and sorry for the delay in responding to it.
